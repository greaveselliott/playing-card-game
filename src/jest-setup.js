// Make Enzyme functions available in all test files without importing
import { shallow, render, mount } from 'enzyme';
import React from 'react';

global.shallow = shallow;
global.render = render;
global.mount = mount;

// Fail tests on any warning
console.error = message => {
    throw new Error(message);
};

/**
 * Function to mock an Enzyme component.
 * @param enzymeRender {function} render function.
 * @param Component {JSXComponent} component type to mock.
 * @param componentProps {Object} optional component properties.
 * @param reduxStoreContext {Object} optional redux store context.
 * @returns Enzyme Component.
 */
global.mockComponent = function (enzymeRender, Component, componentProps = {}, reduxStoreContext = {}) {

    let jsxElement = <Component {...componentProps} />;
    const enzymeComponent = enzymeRender(jsxElement, reduxStoreContext);

    return {
        componentProps,
        enzymeComponent
    }
};

/**
 * Function to create data driven tests against poly component configurations
 * @param Component {JSXComponent} component type to mock.
 * @param configurations {Array of objects} An object containing the test configuration for each test case. Each object within this array must have a 'description' and 'props' property.
 * @returns nothing
 */
global.assertComponentProperties = function (Component, configurations) {

    const expectedProperties = ['description', 'props'];

    configurations.forEach( (testCase) => {

        expectedProperties.forEach((property) => {
            if(!testCase.hasOwnProperty(property)) {
                throw new Error(`Error: assertComponentProperties 'configuration' property does not have required '${property}' property.`);
            }
        });

        const { enzymeComponent } = mockComponent(shallow, Component, testCase.props);

        it(testCase.description, () => {
            expect(enzymeComponent).toMatchSnapshot();
        });
    });
};

global.mockResponse = function (status, statusText, response) {
    return new window.Response(response, {
        status: status,
        statusText: statusText,
        headers: {
            'Content-type': 'application/json'
        }
    });
};