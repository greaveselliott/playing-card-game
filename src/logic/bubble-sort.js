const bubble_sort = (array, key) => {
    var swapped;
    do {
        swapped = false;
        for (var i=0; i < array.length-1; i++) {

            if (array[i][key] > array[i+1][key]) {
                var cached_index = array[i];
                array[i] = array[i+1];
                array[i+1] = cached_index;
                swapped = true;
            }
        }
    } while (swapped);

    return array;
}

export default bubble_sort;