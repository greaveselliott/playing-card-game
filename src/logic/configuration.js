export const card_suits = ["Clubs", "Hearts", "Diamonds", "Spades"];
export const card_numbers = ["2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q","K","A"];

export const deck_factory = (card_suits, card_numbers) => {
    let deck = [];

    card_suits.forEach((suit) => {
        card_numbers.forEach((number) => {
            deck.push({
                id: deck.length,
                suit,
                number
            })
        });
    });

    return deck;
};