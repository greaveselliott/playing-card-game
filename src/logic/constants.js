export const DRAW_INCREMENT = "DRAW_INCREMENT";
export const DRAW_DECREMENT = "DRAW_DECREMENT";
export const DRAW_CARDS = "DRAW_CARDS";
export const SORT_CARDS = "SORT_CARDS";
export const SHUFFLE_DECK = "SHUFFLE_DECK";