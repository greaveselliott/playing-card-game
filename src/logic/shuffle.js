const shuffle = (array) => {
  var copy = [], n = array.length, i;

  while (n) {
    i = Math.floor(Math.random() * n--);
    
    copy.push(array.splice(i, 1)[0]);
  }

  return copy;
}

export default shuffle;