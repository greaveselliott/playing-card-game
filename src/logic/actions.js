import * as CONSTANTS from './constants';

export const draw_increment = () => {
    return {
        type: CONSTANTS.DRAW_INCREMENT
    }
};

export const draw_decrement = () => {
    return {
        type: CONSTANTS.DRAW_DECREMENT
    }
};

export const draw_cards = () => {
    return {
        type: CONSTANTS.DRAW_CARDS
    }
};

export const sort_cards = () => {
    return {
        type: CONSTANTS.SORT_CARDS
    }
};

export const shuffle_deck = () => {
    return {
        type: CONSTANTS.SHUFFLE_DECK
    }
};

export const shuffle_hand = () => {
    return {
        type: CONSTANTS.SHUFFLE_HAND
    }
};