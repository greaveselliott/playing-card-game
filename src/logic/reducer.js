import * as CONSTANTS from './constants';
import _ from 'lodash';
import * as configuration from './configuration';

export const initial_state = {

};

const reducers = (state = initial_state, action) => {
  switch (_.get(action, 'type', undefined)) {


    default:
      return state;
  }
};

export default reducers;
