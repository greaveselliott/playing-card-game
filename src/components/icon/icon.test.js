import React from 'react';
import Icon from './icon';
import '../../jest-setup';
import { card_suits } from '../../logic/configuration';

describe("Icon component", () => {  

    card_suits.forEach((suit) => {
        it("Should render a Icon component.", () => {
            const props = {
                icon: suit
            };
    
            const { enzymeComponent } = mockComponent(shallow, Icon, props);   
    
            expect(enzymeComponent).toMatchSnapshot();
        });
    });
});