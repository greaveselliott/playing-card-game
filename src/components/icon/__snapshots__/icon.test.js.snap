// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon component Should render a Icon component. 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <img
    alt="Clubs"
    className={undefined}
    src="clubs.svg"
/>,
  "nodes": Array [
    <img
      alt="Clubs"
      className={undefined}
      src="clubs.svg"
/>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <Icon
        icon="Clubs"
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "icon": "Clubs",
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <img
          alt="Clubs"
          className={undefined}
          src="clubs.svg"
/>,
        "_debugID": 2,
        "_renderedOutput": <img
          alt="Clubs"
          className={undefined}
          src="clubs.svg"
/>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Icon
    icon="Clubs"
/>,
}
`;

exports[`Icon component Should render a Icon component. 2`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <img
    alt="Spades"
    className={undefined}
    src="spades.svg"
/>,
  "nodes": Array [
    <img
      alt="Spades"
      className={undefined}
      src="spades.svg"
/>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <Icon
        icon="Spades"
/>,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "icon": "Spades",
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <img
          alt="Spades"
          className={undefined}
          src="spades.svg"
/>,
        "_debugID": 4,
        "_renderedOutput": <img
          alt="Spades"
          className={undefined}
          src="spades.svg"
/>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Icon
    icon="Spades"
/>,
}
`;

exports[`Icon component Should render a Icon component. 3`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <img
    alt="Hearts"
    className={undefined}
    src="hearts.svg"
/>,
  "nodes": Array [
    <img
      alt="Hearts"
      className={undefined}
      src="hearts.svg"
/>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <Icon
        icon="Hearts"
/>,
      "_debugID": 5,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "icon": "Hearts",
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 3,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <img
          alt="Hearts"
          className={undefined}
          src="hearts.svg"
/>,
        "_debugID": 6,
        "_renderedOutput": <img
          alt="Hearts"
          className={undefined}
          src="hearts.svg"
/>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Icon
    icon="Hearts"
/>,
}
`;

exports[`Icon component Should render a Icon component. 4`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <img
    alt="Diamonds"
    className={undefined}
    src="diamond.svg"
/>,
  "nodes": Array [
    <img
      alt="Diamonds"
      className={undefined}
      src="diamond.svg"
/>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <Icon
        icon="Diamonds"
/>,
      "_debugID": 7,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "icon": "Diamonds",
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 4,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <img
          alt="Diamonds"
          className={undefined}
          src="diamond.svg"
/>,
        "_debugID": 8,
        "_renderedOutput": <img
          alt="Diamonds"
          className={undefined}
          src="diamond.svg"
/>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Icon
    icon="Diamonds"
/>,
}
`;
